### PR TITLE
 SQL: Refactor args verification of In & conditionals

### PR DESCRIPTION
**Pull request by matriv**
Move verification of arguments for Conditional functions and In
from `Verifier` to the `resolveType()` method of the functions.

Will be merged on top of: #40909

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistryTests.java [hunk 1]
```diff
@@ -41,12 +40,12 @@ public void testNoArgFunction() {
         assertEquals(ur.source(), ur.buildResolved(randomConfiguration(), def).source());
 
         // Distinct isn't supported
-        ParsingException e = expectThrows(ParsingException.class, () ->
+        SqlIllegalArgumentException e = expectThrows(SqlIllegalArgumentException.class, () ->
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistryTests.java]: Why is the `ParsingException` not thrown anymore?

**Review by costin — APPROVED** [review]
LGTM - I'm curious though why the ParsingException is not thrown anymore (and whether the exception hierarchy won't have side-effects).

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistryTests.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistryTests.java]: Thanks for catching that! the default `def()` method to register the functions is catching `IllegalArgumentException` and throwing `ParsingException`, so now some important info (like line number/column & function name is missing). Fixing that.

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/StringUtils.java [hunk 2]
```diff
@@ -23,12 +23,16 @@
...
-public abstract class StringUtils {
+public final class StringUtils {
+
+    private StringUtils() {}
 
     public static final String EMPTY = "";
     public static final String NEW_LINE = "\n";
     public static final String SQL_WILDCARD = "%";
 
+    private static final String[] INTEGER_ORDINALS = new String[] { "th", "st", "nd", "rd", "th", "th", "th", "th", "th", "th" };
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/StringUtils.java]: What's the first `th` for? 10th, 20th, 30th...?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 3]
```diff
@@ -420,46 +420,16 @@ public void testHavingOnScalar() {
...
-    }
-
-    public void testInWithDifferentDataTypesFromLeftValue_SelectClause() {
-        assertEquals("1:14: expected data type [integer], value provided is of type [keyword]",
+    public void testInWithDifferentDataTypesFromLeftValue() {
+        assertEquals("1:8: 1st argument of [1 IN ('foo', 'bar')] must be [integer], found value ['foo'] type [keyword]",
             error("SELECT 1 IN ('foo', 'bar')"));
     }
 
-    public void testInNestedWithDifferentDataTypesFromLeftValue_SelectClause() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: Why removing all these tests?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 4]
```diff
@@ -420,46 +420,16 @@ public void testHavingOnScalar() {
                 error("SELECT int FROM test GROUP BY int HAVING 2 < ABS(int)"));
     }
 
-    public void testInWithDifferentDataTypes_SelectClause() {
-        assertEquals("1:17: expected data type [integer], value provided is of type [keyword]",
+    public void testInWithDifferentDataTypes() {
+        assertEquals("1:8: 2nd argument of [1 IN (2, '3', 4)] must be [integer], found value ['3'] type [keyword]",
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: If there will be a query with `IN` where there are 100 values in its list, won't the error message be hard to follow?

**Review by astefan — APPROVED** [review]
LGTM. Left few minor comments.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 3 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: Because this error was caught in the `Verifier` previously, by traversing the plan tree.
Now the check has been moved to the `In` function itself so there's no need to test ORDER BY, WHERE, etc...

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: Could be, yes, but currently this how we address those type of errors for all functions and I opted of consistency. I think though that now it says: `found value ['3']` for example which the user can search within the IN list and find it quickly.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/StringUtils.java (hunk 2 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/StringUtils.java]: yep
